## Practice Exercise

In [84]:
#To upload the file from the local drive
from google.colab import files
uploaded=files.upload()

Saving Movies.csv to Movies.csv


In this assignment, you will try to find some interesting insights into a few movies released between 1916 and 2016, using Python. You will have to download a movie dataset, write Python code to explore the data, gain insights into the movies, actors, directors, and collections, and submit the code.

#### Some tips before starting the assignment

1. Identify the task to be performed correctly, and only then proceed to write the required code. Don’t perform any incorrect analysis or look for information that isn’t required for the assignment.
2. In some cases, the variable names have already been assigned, and you just need to write code against them. In other cases, the names to be given are mentioned in the instructions. We strongly advise you to use the mentioned names only.
3. Always keep inspecting your data frame after you have performed a particular set of operations.
4. There are some checkpoints given in the IPython notebook provided. They're just useful pieces of information you can use to check if the result you have obtained after performing a particular task is correct or not.
5. Note that you will be asked to refer to documentation for solving some of the questions. That is done on purpose for you to learn new commands and also how to use the documentation.

In [ ]:
# Import the numpy and pandas packages

import numpy as np
import pandas as pd

### Task 1: Reading and Inspection

**Subtask 1.1: Import and read**

Import and read the movie database. Store it in a variable called `movies`.

In [85]:
# Import the numpy and pandas packages
import numpy as np
import pandas as pd
# Write your code for importing the csv file here
movies = pd.read_csv('Movies.csv')
movies


color      director_name  num_critic_for_reviews  duration  \
0     Color      James Cameron                   723.0     178.0   
1     Color     Gore Verbinski                   302.0     169.0   
2     Color         Sam Mendes                   602.0     148.0   
3     Color  Christopher Nolan                   813.0     164.0   
4     Color     Andrew Stanton                   462.0     132.0   
...     ...                ...                     ...       ...   
3848  Color      Shane Carruth                   143.0      77.0   
3849  Color   Neill Dela Llana                    35.0      80.0   
3850  Color   Robert Rodriguez                    56.0      81.0   
3851  Color       Edward Burns                    14.0      95.0   
3852  Color           Jon Gunn                    43.0      90.0   

      director_facebook_likes  actor_3_facebook_likes        actor_2_name  \
0                         0.0                   855.0    Joel David Moore   
1                       563.0                  1000.0       Orlando Bloom   
2                         0.0                   161.0        Rory Kinnear   
3                     22000.0                 23000.0      Christian Bale   
4                       475.0                   530.0     Samantha Morton   
...                       ...                     ...                 ...   
3848                    291.0                     8.0      David Sullivan   
3849                      0.0                     0.0     Edgar Tancangco   
3850                      0.0                     6.0     Peter Marquardt   
3851                      0.0                   133.0  Caitlin FitzGerald   
3852                     16.0                    16.0    Brian Herzlinger   

      actor_1_facebook_likes        gross  \
0                     1000.0  760505847.0   
1                    40000.0  309404152.0   
2                    11000.0  200074175.0   
3                    27000.0  448130642.0   
4                      640.0   73058679.0   
...                      ...          ...   
3848                   291.0     424760.0   
3849                     0.0      70071.0   
3850                   121.0    2040920.0   
3851                   296.0       4584.0   
3852                    86.0      85222.0   

                                   genres  ... num_user_for_reviews language  \
0         Action|Adventure|Fantasy|Sci-Fi  ...               3054.0  English   
1                Action|Adventure|Fantasy  ...               1238.0  English   
2               Action|Adventure|Thriller  ...                994.0  English   
3                         Action|Thriller  ...               2701.0  English   
4                 Action|Adventure|Sci-Fi  ...                738.0  English   
...                                   ...  ...                  ...      ...   
3848                Drama|Sci-Fi|Thriller  ...                371.0  English   
3849                             Thriller  ...                 35.0  English   
3850  Action|Crime|Drama|Romance|Thriller  ...                130.0  Spanish   
3851                         Comedy|Drama  ...                 14.0  English   
3852                          Documentary  ...                 84.0  English   

          country  content_rating       budget  title_year  \
0             USA           PG-13  237000000.0      2009.0   
1             USA           PG-13  300000000.0      2007.0   
2              UK           PG-13  245000000.0      2015.0   
3             USA           PG-13  250000000.0      2012.0   
4             USA           PG-13  263700000.0      2012.0   
...           ...             ...          ...         ...   
3848          USA           PG-13       7000.0      2004.0   
3849  Philippines       Not Rated       7000.0      2005.0   
3850          USA               R       7000.0      1992.0   
3851          USA       Not Rated       9000.0      2011.0   
3852          USA              PG       1100.0      2004.0   

     actor_2_face

**Subtask 1.2: Inspect the dataframe**

Inspect the dataframe's columns, shapes, variable types etc.

In [86]:
# Write your code for inspection here
# Import the numpy and pandas packages
import numpy as np
import pandas as pd
#To check rows and columns are present in the dataframe
movies.shape

#To check nullvalues
#reading the csv file
df = pd.read_csv('Movies.csv')
#obtaining the number of rows
print("number of rows : ", df.shape[0])

# obtaining the number of columns
print("number of columns : ", df.shape[1])
print(" ")

#obtaining null values in each column
column_nan_count = df.isnull().sum()
print("NaN count per column:")
print(column_nan_count)
print("total number of null values:",sum(column_nan_count[column_nan_count>0]))
print(" ")

#obtaining only the coumns which are having null values
print("Columns with null Values")
print(column_nan_count[column_nan_count>0])
print(" ")

#obtaining the total columns having null values
Total_null_Columns=len(column_nan_count[column_nan_count>0])
print("Total Number of coumns with null values are:",Total_null_Columns)










number of rows :  3853
number of columns :  28
 
NaN count per column:
color                         2
director_name                 0
num_critic_for_reviews        1
duration                      1
director_facebook_likes       0
actor_3_facebook_likes        6
actor_2_name                  1
actor_1_facebook_likes        0
gross                         0
genres                        0
actor_1_name                  0
movie_title                   0
num_voted_users               0
cast_total_facebook_likes     0
actor_3_name                  6
facenumber_in_poster          6
plot_keywords                30
movie_imdb_link               0
num_user_for_reviews          0
language                      4
country                       0
content_rating               48
budget                        0
title_year                    0
actor_2_facebook_likes        1
imdb_score                    0
aspect_ratio                 72
movie_facebook_likes          0
dtype: int64
total number of null

#### <font color='red'>Question 1: How many rows and columns are present in the dataframe? </font>
-  <font color='red'>(3821, 26)</font>
-  <font color='red'>(3879, 28)</font>
-  <font color='red'>(3853, 28)</font>
-  <font color='red'>(3866, 26)</font>

#### <font color='red'>Question 2: How many columns have null values present in them? Try writing a code for this instead of counting them manually.</font>

-  <font color='red'>3</font>
-  <font color='red'>6</font>
-  <font color='red'>9</font>
-  <font color='red'>12</font>

### Task 2: Cleaning the Data

**Subtask 2.1: Drop unecessary columns**

For this assignment, you will mostly be analyzing the movies with respect to the ratings, gross collection, popularity of movies, etc. So many of the columns in this dataframe are not required. So it is advised to drop the following columns.
-  color
-  director_facebook_likes
-  actor_1_facebook_likes
-  actor_2_facebook_likes
-  actor_3_facebook_likes
-  actor_2_name
-  cast_total_facebook_likes
-  actor_3_name
-  duration
-  facenumber_in_poster
-  content_rating
-  country
-  movie_imdb_link
-  aspect_ratio
-  plot_keywords

In [91]:
# Check the 'drop' function in the Pandas library - dataframe.drop(list_of_unnecessary_columns, axis = )
# Write your code for dropping the columns here. It is advised to keep inspecting the dataframe after each set of operations
import numpy as np
import pandas as pd
df = pd.read_csv('Movies.csv')
#obtaining the number of rows
print("number of rows before dropping: ", df.shape[0])

# obtaining the number of columns
print("number of columns before dropping: ", df.shape[1])
print(" ")
#dropping the column color
df.drop(['color'],axis=1,inplace=True)
#obtaining the number of rows
print("number of rows after dropping color column: ", df.shape[0])
# obtaining the number of columns
print("number of columns dropping color column: ", df.shape[1])
print(" ")

#Dropping the columns which are not required
print("The director_facebook_likes column is dropped")
df.drop('director_facebook_likes',axis=1,inplace=True)
#obtaining the number of rows
print("number of rows : ", df.shape[0])

# obtaining the number of columns
print("number of columns : ", df.shape[1])
print(" ")

#Dropping the columns which are not required
print("The actor_1_facebook_likes  column is dropped")
df.drop('actor_1_facebook_likes',axis=1,inplace=True)
#obtaining the number of rows
print("number of rows : ", df.shape[0])

# obtaining the number of columns
print("number of columns : ", df.shape[1])
print(" ")

# Drop one or more columns using df.drop()
columns_to_drop = ['actor_2_facebook_likes' ,'actor_3_facebook_likes' ,'actor_2_name' ,'cast_total_facebook_likes' ,'actor_3_name' ,'duration' ,'facenumber_in_poster' ,'content_rating' ,'country' ,'movie_imdb_link' ,'aspect_ratio', 'plot_keywords']
df_dropped = df.drop(columns=columns_to_drop, axis=1)
print("The rows and columns after dropping the multiple unnecessary columns are:")
#obtaining the number of rows
print("number of rows : ", df_dropped.shape[0])
# obtaining the number of columns
print("number of columns : ", df_dropped.shape[1])
print("   ")
print("================The New data after dropping the columns=============================")
df_dropped.head()







number of rows before dropping:  3853
number of columns before dropping:  28
 
number of rows after dropping color column:  3853
number of columns dropping color column:  27
 
The director_facebook_likes column is dropped
number of rows :  3853
number of columns :  26
 
The actor_1_facebook_likes  column is dropped
number of rows :  3853
number of columns :  25
 
The rows and columns after dropping the multiple unnecessary columns are:
number of rows :  3853
number of columns :  13
   
================The New data after dropping the columns=============================


director_name  num_critic_for_reviews        gross  \
0      James Cameron                   723.0  760505847.0   
1     Gore Verbinski                   302.0  309404152.0   
2         Sam Mendes                   602.0  200074175.0   
3  Christopher Nolan                   813.0  448130642.0   
4     Andrew Stanton                   462.0   73058679.0   

                            genres     actor_1_name  \
0  Action|Adventure|Fantasy|Sci-Fi      CCH Pounder   
1         Action|Adventure|Fantasy      Johnny Depp   
2        Action|Adventure|Thriller  Christoph Waltz   
3                  Action|Thriller        Tom Hardy   
4          Action|Adventure|Sci-Fi     Daryl Sabara   

                                 movie_title  num_voted_users  \
0                                    Avatar            886204   
1  Pirates of the Caribbean: At World's End            471220   
2                                   Spectre            275868   
3                     The Dark Knight Rises           1144337   
4                               John Carter            212204   

   num_user_for_reviews language       budget  title_year  imdb_score  \
0                3054.0  English  237000000.0      2009.0         7.9   
1                1238.0  English  300000000.0      2007.0         7.1   
2                 994.0  English  245000000.0      2015.0         6.8   
3                2701.0  English  250000000.0      2012.0         8.5   
4                 738.0  English  263700000.0      2012.0         6.6   

   movie_facebook_likes  
0                 33000  
1                     0  
2                 85000  
3                164000  
4                 24000

#### <font color='red'>Question 3: What is the count of columns in the new dataframe? </font>
-  <font color='red'>10</font>
-  <font color='red'>13</font>
-  <font color='red'>15</font>
-  <font color='red'>17</font>

**Subtask 2.2: Inspect Null values**

As you have seen above, there are null values in multiple columns of the dataframe 'movies'. Find out the percentage of null values in each column of the dataframe 'movies'.

In [133]:
# Check the 'drop' function in the Pandas library - dataframe.drop(list_of_unnecessary_columns, axis = )
# Write your code for dropping the columns here. It is advised to keep inspecting the dataframe after each set of operations
# Check the 'drop' function in the Pandas library - dataframe.drop(list_of_unnecessary_columns, axis = )
# Write your code for dropping the columns here. It is advised to keep inspecting the dataframe after each set of operations
import numpy as np
import pandas as pd
df = pd.read_csv('Movies.csv')
#obtaining the number of rows
print("number of rows before dropping: ", df.shape[0])

# obtaining the number of columns
print("number of columns before dropping: ", df.shape[1])
print(" ")
#dropping the column color
df.drop(['color'],axis=1,inplace=True)
#obtaining the number of rows
print("number of rows after dropping color column: ", df.shape[0])
# obtaining the number of columns
print("number of columns dropping color column: ", df.shape[1])
print(" ")

#Dropping the columns which are not required
print("The director_facebook_likes column is dropped")
df.drop('director_facebook_likes',axis=1,inplace=True)
#obtaining the number of rows
print("number of rows : ", df.shape[0])

# obtaining the number of columns
print("number of columns : ", df.shape[1])
print(" ")

#Dropping the columns which are not required
print("The actor_1_facebook_likes  column is dropped")
df.drop('actor_1_facebook_likes',axis=1,inplace=True)
#obtaining the number of rows
print("number of rows : ", df.shape[0])

# obtaining the number of columns
print("number of columns : ", df.shape[1])
print(" ")

# Drop one or more columns using df.drop()
columns_to_drop = ['actor_2_facebook_likes' ,'actor_3_facebook_likes' ,'actor_2_name' ,'cast_total_facebook_likes' ,'actor_3_name' ,'duration' ,'facenumber_in_poster' ,'content_rating' ,'country' ,'movie_imdb_link' ,'aspect_ratio', 'plot_keywords']
df_dropped = df.drop(columns=columns_to_drop, axis=1)
print("The rows and columns after dropping the multiple unnecessary columns are:")
#obtaining the number of rows
print("number of rows : ", df_dropped.shape[0])
# obtaining the number of columns
print("number of columns : ", df_dropped.shape[1])
print("   ")
#to make a new csv file without changing the original csv
df_dropped.to_csv('new_movies.csv', index=False)

#to obtain a null values from the new movies csv file
df_new = pd.read_csv('new_movies.csv')
column_nan_count = df_new.isnull().sum()
print("========================Null count per column with a new data after dropping the unnecessary columns===================================")

print(column_nan_count)
print(" ")

#obtaining only the coumns which are having null values
print("---------------Columns with null Values-----------------")
print(column_nan_count[column_nan_count>0])
print(" ")

#obtaining the total columns having null values
Total_null_Columns=len(column_nan_count[column_nan_count>0])
print("Total Number of columns with null values are:",Total_null_Columns)
print(" ")

#obtaining the percentage of the columns having null values in new data (after dropping the columns)
Percentage_null_values=df_new.isnull().sum() / df_new.shape[0] * 100.00
print(Percentage_null_values)
print(" ")

#Obtaining the coulmn with highest percentage of null values
Highest_Percentage_null_column=Percentage_null_values.idxmax()
print("----------The Column having highest percentage of null values is: ")
print(Highest_Percentage_null_column)
print("With percentage of null values as:")
print(Percentage_null_values[Highest_Percentage_null_column])





















number of rows before dropping:  3853
number of columns before dropping:  28
 
number of rows after dropping color column:  3853
number of columns dropping color column:  27
 
The director_facebook_likes column is dropped
number of rows :  3853
number of columns :  26
 
The actor_1_facebook_likes  column is dropped
number of rows :  3853
number of columns :  25
 
The rows and columns after dropping the multiple unnecessary columns are:
number of rows :  3853
number of columns :  13
   
========================Null count per column with a new data after dropping the unnecessary columns===================================
director_name             0
num_critic_for_reviews    1
gross                     0
genres                    0
actor_1_name              0
movie_title               0
num_voted_users           0
num_user_for_reviews      0
language                  4
budget                    0
title_year                0
imdb_score                0
movie_facebook_likes      0
dtype: in

#### <font color='red'>Question 4: Which column has the highest percentage of null values? </font>
-  <font color='red'>language</font>
-  <font color='red'>genres</font>
-  <font color='red'>num_critic_for_reviews</font>
-  <font color='red'>imdb_score</font>

**Subtask 2.3: Fill NaN values**

You might notice that the `language` column has some NaN values. Here, on inspection, you will see that it is safe to replace all the missing values with `'English'`.

In [181]:
# Write your code for filling the NaN values in the 'language' column here
# Check the 'drop' function in the Pandas library - dataframe.drop(list_of_unnecessary_columns, axis = )
# Write your code for dropping the columns here. It is advised to keep inspecting the dataframe after each set of operations
# Check the 'drop' function in the Pandas library - dataframe.drop(list_of_unnecessary_columns, axis = )
# Write your code for dropping the columns here. It is advised to keep inspecting the dataframe after each set of operations
import numpy as np
import pandas as pd
df = pd.read_csv('Movies.csv')
#obtaining the number of rows
print("number of rows before dropping: ", df.shape[0])

# obtaining the number of columns
print("number of columns before dropping: ", df.shape[1])
print(" ")
#dropping the column color
df.drop(['color'],axis=1,inplace=True)
#obtaining the number of rows
print("number of rows after dropping color column: ", df.shape[0])
# obtaining the number of columns
print("number of columns dropping color column: ", df.shape[1])
print(" ")

#Dropping the columns which are not required
print("The director_facebook_likes column is dropped")
df.drop('director_facebook_likes',axis=1,inplace=True)
#obtaining the number of rows
print("number of rows : ", df.shape[0])

# obtaining the number of columns
print("number of columns : ", df.shape[1])
print(" ")

#Dropping the columns which are not required
print("The actor_1_facebook_likes  column is dropped")
df.drop('actor_1_facebook_likes',axis=1,inplace=True)
#obtaining the number of rows
print("number of rows : ", df.shape[0])

# obtaining the number of columns
print("number of columns : ", df.shape[1])
print(" ")

# Drop one or more columns using df.drop()
columns_to_drop = ['actor_2_facebook_likes' ,'actor_3_facebook_likes' ,'actor_2_name' ,'cast_total_facebook_likes' ,'actor_3_name' ,'duration' ,'facenumber_in_poster' ,'content_rating' ,'country' ,'movie_imdb_link' ,'aspect_ratio', 'plot_keywords']
df_dropped = df.drop(columns=columns_to_drop, axis=1)
print("The rows and columns after dropping the multiple unnecessary columns are:")
#obtaining the number of rows
print("number of rows : ", df_dropped.shape[0])
# obtaining the number of columns
print("number of columns : ", df_dropped.shape[1])
print("   ")
#to make a new csv file without changing the original csv
df_dropped.to_csv('new_movies.csv', index=False)

#to obtain a null values from the new movies csv file
df_new = pd.read_csv('new_movies.csv')
column_nan_count = df_new.isnull().sum()
print("========================Null count per column with a new data after dropping the unnecessary columns===================================")

print(column_nan_count)
print(" ")

#obtaining only the coumns which are having null values
print("---------------Columns with null Values-----------------")
print(column_nan_count[column_nan_count>0])
print(" ")

#obtaining the total columns having null values
Total_null_Columns=len(column_nan_count[column_nan_count>0])
print("Total Number of columns with null values are:",Total_null_Columns)
print(" ")

#obtaining the percentage of the columns having null values in new data (after dropping the columns)
Percentage_null_values=df_new.isnull().sum() / df_new.shape[0] * 100.00
print(Percentage_null_values)
print(" ")

#Obtaining the coulmn with highest percentage of null values
Highest_Percentage_null_column=Percentage_null_values.idxmax()
print("----------The Column having highest percentage of null values is: ")
print(Highest_Percentage_null_column)
print("With percentage of null values as:")
print(Percentage_null_values[Highest_Percentage_null_column])
print(" ")

#Printing the language column
print(df_new['language'])
print(" ")
print("the count of the language before replacing the null values")
print(df_new['language'].value_counts())
print(" ")

#to get the uniques values of the column lanuage
print("The unique languages in the column language is as like below\n",df_new.language.unique())
print(" ")
print("======================================================")
print(" ")

#the language column has some NaN values,replacing all the missing values with 'English'.
df_new.language = df_new.language.fillna('English')
print(df_new['language'])
print(" ")
print(" ")
print("=========================")

# Counting NaN values in the 'language' column
language_column_nan_count = df_new['language'].isna().sum()
print("Number of Null values in the 'language' column after replacing the null values with English:", language_column_nan_count)
print("================================================================================================")
print("================================================================================================")
print("the count of the language after  replacing the null values")
print(df_new['language'].value_counts())
print(" ")
print("======================================================")
print(" ")


#to get a count of english language mpvies after replacing the null values
count_of_english_movies = df_new['language'].value_counts()['English']
print(f"The number of English movies  is: {count_of_english_movies }")











number of rows before dropping:  3853
number of columns before dropping:  28
 
number of rows after dropping color column:  3853
number of columns dropping color column:  27
 
The director_facebook_likes column is dropped
number of rows :  3853
number of columns :  26
 
The actor_1_facebook_likes  column is dropped
number of rows :  3853
number of columns :  25
 
The rows and columns after dropping the multiple unnecessary columns are:
number of rows :  3853
number of columns :  13
   
========================Null count per column with a new data after dropping the unnecessary columns===================================
director_name             0
num_critic_for_reviews    1
gross                     0
genres                    0
actor_1_name              0
movie_title               0
num_voted_users           0
num_user_for_reviews      0
language                  4
budget                    0
title_year                0
imdb_score                0
movie_facebook_likes      0
dtype: in

#### <font color='red'>Question 5: What is the count of movies made in English language after replacing the NaN values with English? </font>
-  <font color='red'>3670</font>
-  <font color='red'>3674</font>
-  <font color='red'>3668</font>
-  <font color='red'>3672</font>

### Task 3: Data Analysis

**Subtask 3.1: Change the unit of columns**

Convert the unit of the `budget` and `gross` columns from `$` to `million $`.

In [263]:
# Write your code for unit conversion here
import numpy as np
import pandas as pd
df = pd.read_csv('Movies.csv')
#obtaining the number of rows
print("number of rows before dropping: ", df.shape[0])

# obtaining the number of columns
print("number of columns before dropping: ", df.shape[1])
print(" ")
#dropping the column color
df.drop(['color'],axis=1,inplace=True)
#obtaining the number of rows
print("number of rows after dropping color column: ", df.shape[0])
# obtaining the number of columns
print("number of columns dropping color column: ", df.shape[1])
print(" ")

#Dropping the columns which are not required
print("The director_facebook_likes column is dropped")
df.drop('director_facebook_likes',axis=1,inplace=True)
#obtaining the number of rows
print("number of rows : ", df.shape[0])

# obtaining the number of columns
print("number of columns : ", df.shape[1])
print(" ")

#Dropping the columns which are not required
print("The actor_1_facebook_likes  column is dropped")
df.drop('actor_1_facebook_likes',axis=1,inplace=True)
#obtaining the number of rows
print("number of rows : ", df.shape[0])

# obtaining the number of columns
print("number of columns : ", df.shape[1])
print(" ")

# Drop one or more columns using df.drop()
columns_to_drop = ['actor_2_facebook_likes' ,'actor_3_facebook_likes' ,'actor_2_name' ,'cast_total_facebook_likes' ,'actor_3_name' ,'duration' ,'facenumber_in_poster' ,'content_rating' ,'country' ,'movie_imdb_link' ,'aspect_ratio', 'plot_keywords']
df_dropped = df.drop(columns=columns_to_drop, axis=1)
print("The rows and columns after dropping the multiple unnecessary columns are:")
#obtaining the number of rows
print("number of rows : ", df_dropped.shape[0])
# obtaining the number of columns
print("number of columns : ", df_dropped.shape[1])
print("   ")
#to make a new csv file without changing the original csv
df_dropped.to_csv('new_movies.csv', index=False)
#printing the new data columns
df_new = pd.read_csv('new_movies.csv')
column_names = df_new.columns
print("The columns exist in the new data are")
print(column_names)
print(" ")
#printing the data of the gross and buget column
print(df_new['gross'], df_new['budget'])
print("===================================================================================")
print("===================================================================================")

#converting the gross and budget from $ to Million$
df_new['gross'] = df_new['gross'] / 1000000
df_new['budget'] = df_new['budget'] / 1000000
print(" ")
print("======================After converting the gross and budget from $ to Million$===================")
print(df_new['gross'], df_new['budget'])



number of rows before dropping:  3853
number of columns before dropping:  28
 
number of rows after dropping color column:  3853
number of columns dropping color column:  27
 
The director_facebook_likes column is dropped
number of rows :  3853
number of columns :  26
 
The actor_1_facebook_likes  column is dropped
number of rows :  3853
number of columns :  25
 
The rows and columns after dropping the multiple unnecessary columns are:
number of rows :  3853
number of columns :  13
   
The columns exist in the new data are
Index(['director_name', 'num_critic_for_reviews', 'gross', 'genres',
       'actor_1_name', 'movie_title', 'num_voted_users',
       'num_user_for_reviews', 'language', 'budget', 'title_year',
       'imdb_score', 'movie_facebook_likes'],
      dtype='object')
 
0       760505847.0
1       309404152.0
2       200074175.0
3       448130642.0
4        73058679.0
           ...     
3848       424760.0
3849        70071.0
3850      2040920.0
3851         4584.0
3852    

**Subtask 3.2: Find the movies with highest profit**

   1. Create a new column called `profit` which contains the difference of the two columns: `gross` and `budget`.
   2. Sort the dataframe using the `profit` column as reference. (Find which command can be used here to sort entries from the documentation)
   3. Extract the top ten profiting movies in descending order and store them in a new dataframe - `top10`

In [264]:
# Write your code for creating the profit column here
# Write your code for unit conversion here
import numpy as np
import pandas as pd
df = pd.read_csv('Movies.csv')
#obtaining the number of rows
print("number of rows before dropping: ", df.shape[0])

# obtaining the number of columns
print("number of columns before dropping: ", df.shape[1])
print(" ")
#dropping the column color
df.drop(['color'],axis=1,inplace=True)
#obtaining the number of rows
print("number of rows after dropping color column: ", df.shape[0])
# obtaining the number of columns
print("number of columns dropping color column: ", df.shape[1])
print(" ")

#Dropping the columns which are not required
print("The director_facebook_likes column is dropped")
df.drop('director_facebook_likes',axis=1,inplace=True)
#obtaining the number of rows
print("number of rows : ", df.shape[0])

# obtaining the number of columns
print("number of columns : ", df.shape[1])
print(" ")

#Dropping the columns which are not required
print("The actor_1_facebook_likes  column is dropped")
df.drop('actor_1_facebook_likes',axis=1,inplace=True)
#obtaining the number of rows
print("number of rows : ", df.shape[0])

# obtaining the number of columns
print("number of columns : ", df.shape[1])
print(" ")

# Drop one or more columns using df.drop()
columns_to_drop = ['actor_2_facebook_likes' ,'actor_3_facebook_likes' ,'actor_2_name' ,'cast_total_facebook_likes' ,'actor_3_name' ,'duration' ,'facenumber_in_poster' ,'content_rating' ,'country' ,'movie_imdb_link' ,'aspect_ratio', 'plot_keywords']
df_dropped = df.drop(columns=columns_to_drop, axis=1)
print("The rows and columns after dropping the multiple unnecessary columns are:")
#obtaining the number of rows
print("number of rows : ", df_dropped.shape[0])
# obtaining the number of columns
print("number of columns : ", df_dropped.shape[1])
print("   ")
#to make a new csv file without changing the original csv
df_dropped.to_csv('new_movies.csv', index=False)
#printing the new data columns
df_new = pd.read_csv('new_movies.csv')
column_names = df_new.columns
print("The columns exist in the new data are")
print(column_names)
print(" ")
print("===================================================================================")
print("===================================================================================")

#converting the gross and budget from $ to Million$
df_new['gross'] = df_new['gross'] / 1000000
df_new['budget'] = df_new['budget'] / 1000000
print(" ")

#difference between gross and budget and adding profit colum
new_column = df_new['gross'] - df_new['budget']
df_new['profit'] = new_column
print("The profit is =====================")
print(df_new['profit'])
print("===================================================================================")
print("==================================================================================")
print("The data after adding the profit column")
print(df_new)
print("===================================================================================")
print("==================================================================================")
After_adding_profit_columns=df_new.columns
print("The columns exist in the new data are")
print(After_adding_profit_columns)






number of rows before dropping:  3853
number of columns before dropping:  28
 
number of rows after dropping color column:  3853
number of columns dropping color column:  27
 
The director_facebook_likes column is dropped
number of rows :  3853
number of columns :  26
 
The actor_1_facebook_likes  column is dropped
number of rows :  3853
number of columns :  25
 
The rows and columns after dropping the multiple unnecessary columns are:
number of rows :  3853
number of columns :  13
   
The columns exist in the new data are
Index(['director_name', 'num_critic_for_reviews', 'gross', 'genres',
       'actor_1_name', 'movie_title', 'num_voted_users',
       'num_user_for_reviews', 'language', 'budget', 'title_year',
       'imdb_score', 'movie_facebook_likes'],
      dtype='object')
 
 
The profit is =====================
0       523.505847
1         9.404152
2       -44.925825
3       198.130642
4      -190.641321
           ...    
3848      0.417760
3849      0.063071
3850      2.033920

In [330]:
# Write your code for sorting the dataframe here
df_new_sorted = df_new.sort_values(by='profit', ascending=False)
print(df_new_sorted)








         director_name  num_critic_for_reviews       gross  \
0        James Cameron                   723.0  760.505847   
28     Colin Trevorrow                   644.0  652.177271   
25       James Cameron                   315.0  658.672302   
2704      George Lucas                   282.0  460.935665   
2748  Steven Spielberg                   215.0  434.949459   
...                ...                     ...         ...   
2147   Katsuhiro Ôtomo                   105.0    0.410388   
2136    Hayao Miyazaki                   174.0    2.298191   
2693      Lajos Koltai                    73.0    0.195888   
3282    Chan-wook Park                   202.0    0.211667   
2678      Joon-ho Bong                   363.0    2.201412   

                                                 genres         actor_1_name  \
0                       Action|Adventure|Fantasy|Sci-Fi          CCH Pounder   
28                     Action|Adventure|Sci-Fi|Thriller  Bryce Dallas Howard   
25             

In [331]:
 # Write your code to get the top 10 profiting movies here
 top10 = df_new_sorted.head(10)
print(top10)
print(" ")
print("The top 5th movie ")
top10.iloc[4]['movie_title']







          director_name  num_critic_for_reviews       gross  \
0         James Cameron                   723.0  760.505847   
28      Colin Trevorrow                   644.0  652.177271   
25        James Cameron                   315.0  658.672302   
2704       George Lucas                   282.0  460.935665   
2748   Steven Spielberg                   215.0  434.949459   
16          Joss Whedon                   703.0  623.279547   
482        Roger Allers                   186.0  422.783777   
230        George Lucas                   320.0  474.544677   
64    Christopher Nolan                   645.0  533.316061   
419           Gary Ross                   673.0  407.999255   

                                        genres         actor_1_name  \
0              Action|Adventure|Fantasy|Sci-Fi          CCH Pounder   
28            Action|Adventure|Sci-Fi|Thriller  Bryce Dallas Howard   
25                               Drama|Romance    Leonardo DiCaprio   
2704           Action|

'E.T. the Extra-Terrestrial\xa0'

**Checkpoint:** You might spot two movies directed by `James Cameron` in the list.

#### <font color='red'>Question 6: Which movie is ranked 5th from the top in the list obtained? </font>
-  <font color='red'>E.T. the Extra-Terrestrial</font>
-  <font color='red'>The Avengers</font>
-  <font color='red'>The Dark Knight</font>
-  <font color='red'>Titanic</font>

**Subtask 3.3: Find IMDb Top 250**

Create a new dataframe `IMDb_Top_250` and store the top 250 movies with the highest IMDb Rating (corresponding to the column: `imdb_score`). Also make sure that for all of these movies, the `num_voted_users` is greater than 25,000.

Also add a `Rank` column containing the values 1 to 250 indicating the ranks of the corresponding films.

In [336]:
# Write your code for extracting the top 250 movies as per the IMDb score here. Make sure that you store it in a new dataframe
# and name that dataframe as 'IMDb_Top_250'
IMDb_Top_250 = df_new.query("num_voted_users > 25000").sort_values('imdb_score',ascending=False).head(250)
IMDb_Top_250 = IMDb_Top_250.assign(Rank = lambda x: range(1, len(x) + 1))
IMDb_Top_250



director_name  num_critic_for_reviews        gross  \
1795        Frank Darabont                   199.0   28341469.0   
3016  Francis Ford Coppola                   208.0  134821952.0   
64       Christopher Nolan                   645.0  533316061.0   
2543  Francis Ford Coppola                   149.0   57300000.0   
325          Peter Jackson                   328.0  377019252.0   
...                    ...                     ...          ...   
67             Jon Favreau                   486.0  318298180.0   
73              Doug Liman                   585.0  100189501.0   
3052           Yash Chopra                    29.0    2921738.0   
95           Peter Jackson                   645.0  303001229.0   
3170  Christophe Barratier                   112.0    3629758.0   

                              genres           actor_1_name  \
1795                     Crime|Drama         Morgan Freeman   
3016                     Crime|Drama              Al Pacino   
64       Action|Crime|Drama|Thriller         Christian Bale   
2543                     Crime|Drama         Robert De Niro   
325   Action|Adventure|Drama|Fantasy          Orlando Bloom   
...                              ...                    ...   
67           Action|Adventure|Sci-Fi      Robert Downey Jr.   
73           Action|Adventure|Sci-Fi             Tom Cruise   
3052           Drama|Musical|Romance         Shah Rukh Khan   
95                 Adventure|Fantasy           Aidan Turner   
3170                     Drama|Music  Jean-Baptiste Maunier   

                                         movie_title  num_voted_users  \
1795                       The Shawshank Redemption           1689764   
3016                                  The Godfather           1155770   
64                                  The Dark Knight           1676169   
2543                         The Godfather: Part II            790926   
325   The Lord of the Rings: The Return of the King           1215718   
...                                              ...              ...   
67                                         Iron Man            696338   
73                                 Edge of Tomorrow            431620   
3052                                     Veer-Zaara             34449   
95                The Hobbit: An Unexpected Journey            637246   
3170                                     The Chorus             44151   

      num_user_for_reviews language       budget  title_year  imdb_score  \
1795                4144.0  English   25000000.0      1994.0         9.3   
3016                2238.0  English    6000000.0      1972.0         9.2   
64                  4667.0  English  185000000.0      2008.0         9.0   
2543                 650.0  English   13000000.0      1974.0         9.0   
325                 3189.0  English   94000000.0      2003.0         8.9   
...                    ...      ...          ...         ...         ...   
67                  1055.0  English  140000000.0      2008.0         7.9   
73                   741.0  English  178000000.0      2014.0         7.9   
3052                 119.0    Hindi    7000000.0      2004.0         7.9   
95                  1367.0  English  180000000.0      2012.0         7.9   
3170                 110.0   French    5500000.0      2004.0         7.9   

      movie_facebook_likes  Rank  
1795                108000     1  
3016                 43000     2  
64                   37000     3  
2543                 14000     4  
325                  16000     5  
...                    ...   ...  
67                   10000   246  
73                   77000   247  
3052                  2000   248  
95                  166000   249  
3170                     0   250  

[250 rows x 14 columns]

#### <font color='red'>Question 7: Suppose movies are divided into 5 buckets based on the IMDb ratings: </font>
-  <font color='red'>7.5 to 8</font>
-  <font color='red'>8 to 8.5</font>
-  <font color='red'>8.5 to 9</font>
-  <font color='red'>9 to 9.5</font>
-  <font color='red'>9.5 to 10</font>

<font color = 'red'> Which bucket holds the maximum number of movies from *IMDb_Top_250*? </font>

In [337]:
#creating bins [7.5 to 8,8 to 8.5,8.5 to 9,9 to 9.5,9.5 to 10]
IMDb_Top_250['rating_bucket'] = pd.cut(IMDb_Top_250['imdb_score'], bins=[7.5, 8, 8.5, 9, 9.5, 10], labels=["7.5-8", "8-8.5", "8.5-9", "9-9.5", "9.5-10"])
IMDb_Top_250['rating_bucket'].value_counts().idxmax()

'8-8.5'

**Subtask 3.4: Find the critic-favorite and audience-favorite actors**

   1. Create three new dataframes namely, `Meryl_Streep`, `Leo_Caprio`, and `Brad_Pitt` which contain the movies in which the actors: 'Meryl Streep', 'Leonardo DiCaprio', and 'Brad Pitt' are the lead actors. Use only the `actor_1_name` column for extraction. Also, make sure that you use the names 'Meryl Streep', 'Leonardo DiCaprio', and 'Brad Pitt' for the said extraction.
   2. Append the rows of all these dataframes and store them in a new dataframe named `Combined`.
   3. Group the combined dataframe using the `actor_1_name` column.
   4. Find the mean of the `num_critic_for_reviews` and `num_user_for_review` and identify the actors which have the highest mean.

In [338]:
# Write your code for creating three new dataframes here
# Include all movies in which Meryl_Streep is the lead
Meryl_Streep = df_new[df_new.actor_1_name == 'Meryl Streep']


In [340]:
# Include all movies in which Leo_Caprio is the lead
Leo_Caprio = df_new[df_new.actor_1_name == 'Leo_Caprio']


In [339]:
 # Include all movies in which Brad_Pitt is the lead
 Brad_Pitt = df_new[df_new.actor_1_name == 'Brad_Pitt']


In [321]:
# Write your code for combining the three dataframes here
Combined = pd.concat([Meryl_Streep, Brad_Pitt, Leo_Caprio])


In [341]:
# Write your code for grouping the combined dataframe here
grouped_data = Combined.groupby('actor_1_name')


In [347]:
# Write the code for finding the mean of critic reviews and audience reviews here
mean_reviews = grouped_data[['num_critic_for_reviews', 'num_user_for_reviews']].mean()


#### <font color='red'>Question 9: Which actor is highest rated among the three actors according to the critics?</font>
-  <font color='red'>Meryl Streep</font>
-  <font color='red'>Leonardo DiCaprio</font>
-  <font color='red'>Brad Pitt</font>

In [345]:
mean_reviews = grouped_data[['num_user_for_reviews']].mean()
mean_reviews


num_user_for_reviews
actor_1_name                      
Meryl Streep            297.181818

#### <font color='red'>Question 8: Which actor is highest rated among the three actors according to the user reviews? </font>
-  <font color='red'>Meryl Streep</font>
-  <font color='red'>Leonardo DiCaprio</font>
-  <font color='red'>Brad Pitt</font>

In [317]:
mean_reviews = grouped_data[['num_critic_for_reviews']].mean()
mean_reviews

num_critic_for_reviews
actor_1_name                        
Meryl Streep              181.454545